Improve integer assumptions, Mod, and prime factorization of factorial

ylemkimon opened the pull request:
**Integer assumptions:**
* positive non-prime even -> composite
* odd composite >= 9
* composite >= 4
* even non-prime >= 4
* Mul and Pow: added _eval_is_composite
* Pow: improved _eval_is_prime: fixes #9151
* factorial: added _eval_is_even: fixes #9150
* factorial: non-negative check for _eval_is_composite

**Mod**
* raise ZeroDivisionError when modulo divisor is zero
* evaluate to zero when dividend is zero
* fixed incorrect result of Mod by difference: fixes #13125
* improved mod denesting
* call _eval_Mod for function-specific implementation of modulo operation
* Pow: evaluate to zero when base % q == 0
* Pow: use built-in pow modular exponentiation
* factorial: evaluate to zero when >=q!: fixes #8677
* factorial: implemented Wilson's theorem and its inverse

**Prime factorization**
* implemented Legendre's formula for prime factorization(factorint) of unevaluated factorial.

ylemkimon committed:
Assume positive non-prime even number composite

A positive non-prime even number is composite number.

ylemkimon committed:
Make _monotonic_sign reflect prime/comp assumption

Smallest odd composite number is 9, composite number is 4, and even
non-prime number is 4.

ylemkimon committed:
Add _eval_is_composite for Mul

Added _eval_is_composite for Mul based on _eval_is_prime.

ylemkimon committed:
Improve is_prime and add is_composite for Pow

* Improved _eval_is_prime: an integer raised to the n(>=2)-th power
cannot be a prime
* Added _eval_is_composite based on _eval_is_prime

ylemkimon committed:
Add _eval_is_even for factorial...

A factorial is even when >=2!. Also added a non-negative check to
_eval_is_composite.

ylemkimon committed:
Raise ZeroDivisionError when modulo by zero

Made Mod raise ZeroDivisionError when modulo divisor is zero.

ylemkimon committed:
Fix incorrect Mod by difference

Fixed incorrect result of Mod by difference.

ylemkimon committed:
Improve Mod denesting

Improved Mod denesting when qinner % q == 0 and |qinner| < |q|.

ylemkimon committed:
Call _eval_Mod in Mod...

* Implemented _eval_Mod for function-specific implementation of modulo
operation.
* Made Mod evaluate to zero when dividend is zero
* Moved Mod evaluation of power to _eval_Mod in power.
* Made Mod evaluate to zero when base % q == 0 in Pow
* Moved and improved Mod Pow test cases

ylemkimon committed:
Use built-in pow modular exponentiation for Pow Mod

Made use built-in pow modular exponentiation for unevaluated Integer Pow
modulo operation.

ylemkimon committed:
Add _eval_Mod for factorial

* Factorial is divisible by n, when >=n!
* Implemented Wilson's theorem and its inverse for unevaluated or
symbolic factorial

ylemkimon committed:
Implement Legendre's formula for factorial factorization

Implemented Legendre's formula for prime factorization(factorint) of
unevaluated factorial.

ylemkimon committed:
Fix test cases

Fixes test cases of which results are altered due to sympy/sympy#13236.

smichr commented on sympy/core/mod.py:
```diff
@@ -36,12 +36,11 @@ def doit(p, q):
-                    p.is_Pow and p.exp.is_integer and p.base == q and q.is_integer
-                    and p.exp.is_positive or
-                    p.is_integer and q == 1):
+            if p == S.Zero or p == q or p == -q or (p.is_integer and q == 1):
```
Is it intentional to use `p == S.Zero` which is true for p in (0, 0.0) and `q == 1` which is true for q in (1, 1.0)?

ylemkimon replied:
@smichr, I just left original `q == 1` comparison and added `p == S.Zero` comparison in similar manner, but I don't think there is a problem.

smichr commented on sympy/core/mul.py:
```diff
@@ -1332,11 +1328,21 @@ def _eval_is_prime(self):
                 r *= arg
```
I believe it is more efficient to do `r = Mul(*self.args)` or simply, return `Mul(*self.args).is_prime` -- the same below in `_eval_is_composite`.

smichr commented on sympy/core/tests/test_assumptions.py:
```diff
@@ -764,32 +767,44 @@ def test_Pow_is_algebraic():
+    assert ( (x+1)*(y+1) ).is_prime is None
+    assert ( (x+1)*(y+1) ).is_composite is None
 
     assert Mul(6, S.Half, evaluate=False).is_prime is True
```
I wonder if the case where any arg is not an integer should simply return None rather than reevaluating the expression. This would be more in keeping with simpler conservative answers rather than trying to do simplification that should be done by the user. My growing feeling is that SymPy tries, too often, to use internal simplification to get a better result but in turn makes things slower because of it. Do you have any thoughts on this?

ylemkimon replied:
@smichr I agree, I think we can remove evaluating unevaluated Mul/Pow part, since unevaluated Mul/Pow is generated only if explicitly passed 'evaluate=False' by user and it means the user doesn't want simplification nor evaluation of its value.

ylemkimon committed:
Remove is_prime/composite for unevaluated Mul/Pow

Removed evaluation of unevaluated Mul/Pow in _eval_is_prime and
_eval_is_composite. See sympy/sympy#13236.

ylemkimon committed:
Trigger Travis build

ylemkimon commented:
Travis slow test failure seems to be Travis error([see](https://travis-ci.org/ylemkimon/sympy/builds/271703060)). Pushed an empty commit to re-run Travis.

smichr commented:
Thanks -- it's in.
